provider/oci/networking.go

```diff
@@ -4,65 +4,981 @@
 package oci
 
 import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
 	"github.com/juju/errors"
+	"github.com/juju/utils/set"
 	"gopkg.in/juju/names.v2"
 
 	"github.com/juju/juju/environs"
-	"github.com/juju/juju/environs/context"
+	envcontext "github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/tags"
 	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+const (
+	// DefaultAddressSpace is the subnet to use for the default juju VCN
+	// An individual subnet will be created from this class, for each
+	// availability domain.
+	DefaultAddressSpace = "10.0.0.0/16"
+
+	SubnetPrefixLength = "24"
+
+	VcnNamePrefix         = "juju-vcn"
+	SecListNamePrefix     = "juju-seclist"
+	SubnetNamePrefix      = "juju-subnet"
+	InternetGatewayPrefix = "juju-ig"
+	RouteTablePrefix      = "juju-rt"
+)
+
+var (
+	tcpProtocolNumber  = "6"
+	udpProtocolNumber  = "17"
+	icmpProtocolNumber = "1"
+	allProtocols       = "all"
 )
 
+// TODO(gsamfira): Use "local" instead? make configurable?
+var DnsLabelTld = "juju"
+
+func (e *Environ) vcnName(controllerUUID string) *string {
+	name := fmt.Sprintf("%s-%s", VcnNamePrefix, controllerUUID)
+	return &name
+}
+
+func (e *Environ) getVCNStatus(vcnID *string) (string, error) {
+	request := ociCore.GetVcnRequest{
+		VcnId: vcnID,
+	}
+
+	response, err := e.Networking.GetVcn(context.Background(), request)
+	if err != nil {
+		if e.isNotFound(response.RawResponse) {
+			return "", errors.NotFoundf("vcn %q", *vcnID)
+		} else {
+			return "", err
+		}
+	}
+	return string(response.Vcn.LifecycleState), nil
+}
+
+func (e *Environ) getVcn(controllerUUID string) (ociCore.Vcn, error) {
+	request := ociCore.ListVcnsRequest{
+		CompartmentId: e.ecfg().compartmentID(),
+	}
+
+	response, err := e.Networking.ListVcns(context.Background(), request)
+	if err != nil {
+		return ociCore.Vcn{}, errors.Trace(err)
+	}
+	name := e.vcnName(controllerUUID)
+
+	if len(response.Items) > 0 {
+		for _, val := range response.Items {
+			// NOTE(gsamfira): Display names are not unique. We only care
+			// about VCNs that have been created for this controller.
+			// An extra check is required besides the name.
+			if *val.DisplayName != *name {
+				continue
+			}
+			if tag, ok := val.FreeformTags[tags.JujuController]; ok {
+				if tag == controllerUUID {
+					return val, nil
+				}
+			}
+		}
+	}
+	return ociCore.Vcn{}, errors.NotFoundf("no such VCN: %s", *name)
+}
+
+func (e *Environ) secListName(controllerUUID string) string {
+	return fmt.Sprintf("juju-seclist-%s", controllerUUID)
+}
+
+func (e *Environ) ensureVCN(controllerUUID string) (ociCore.Vcn, error) {
+	logger.Infof("ensuring that VCN is created")
```

Comment:
I'm not sure this Infof is as useful as just reporting on the result of the query. If the query is known to take a long time, having a "preparing/ensuring" is ok, but the more interesting log is almost always the result of the query, and you can combine those into the final output. So possibly drop this to Debug but make sure we have a good message of "VCN not found, creating", or something to that effect.